backend/infrahub/core/node/resource_manager/ip_address_pool.py

```diff
@@ -62,25 +63,26 @@ async def get_resource(
 
         prefixlen = prefixlen or data.get("prefixlen") or self.default_prefix_length.value  # type: ignore[attr-defined]
 
-        next_address = await self.get_next(db=db, prefixlen=prefixlen)
-
-        target_schema = registry.get_node_schema(name=address_type, branch=branch)
-        node = await Node.init(db=db, schema=target_schema, branch=branch, at=at)
-        try:
-            await node.new(db=db, address=str(next_address), ip_namespace=ip_namespace, **data)
-        except ValidationError as exc:
-            raise ValueError(f"IPAddressPool: {self.name.value} | {exc!s}") from exc  # type: ignore[attr-defined]
-        await node.save(db=db, at=at)
-        reconciler = IpamReconciler(db=db, branch=branch)
-        await reconciler.reconcile(ip_value=next_address, namespace=ip_namespace.id, node_uuid=node.get_id())
-
-        if identifier:
-            query_set = await IPAddressPoolSetReserved.init(
-                db=db, pool_id=self.id, identifier=identifier, address_id=node.id, at=at
-            )
-            await query_set.execute(db=db)
-
-        return node
+        async with lock.registry.get(name=self.get_id(), namespace="resource_pool"):
```

Comment:
I think whole `get_resource()` has to be locked, there might be TOCTOU issues when concurrent mutations occur with an identifier defined? (regarding line 40 to 51)

Reply:
I might be missing something here, can you walk me through the scenario you see? The way I was thinking was that if we'd already had a reservation in the database/pool we wouldn't need to lock the pool to get that reservation. However, if we don't have a previous reservation, we'd want to have a lock while making the reservation.

Are you thinking if we have two mutations that try to update the same node at the same time and as such have the same `identifier`?

Reply:
> However, if we don't have a previous reservation, we'd want to have a lock while making the reservation.

For that specific statement, I think you need to lock early (before checking for a reservation).

> Are you thinking if we have two mutations that try to update the same node at the same time and as such have the same identifier?

Exactly, to solve this specific issue.

Reply:
I'm not sure either if we'll ever run into this but I moved the lock regardless. I think the main issue might be multiple mutations on against the same node. But better safe than sorry, it might take a slight performance hit but I don't think this in itself will be in the hot path.

Also I messed up this PR as I brought in the wrong branch.. Opened #7025 as a replacement instead.

Reply:
Thanks, looks good.
I'll run additional tests with a branch containing this and https://github.com/opsmill/infrahub/pull/6888 together